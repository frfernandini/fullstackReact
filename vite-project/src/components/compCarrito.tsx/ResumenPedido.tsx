import React from 'react';

interface ResumenPedidoProps {
  subtotal: number;
  abrirCheckout: () => void;
}

export const ResumenPedido: React.FC<ResumenPedidoProps> = ({ subtotal, abrirCheckout }) => (
  <div className="order-summary sticky-top">
    <div className="card">
      <div className="card-body">
        <h4 className="card-title">Resumen del Pedido</h4>
        <hr />
        <div className="d-flex justify-content-between mb-3">
          <span>Subtotal:</span>
          <span id="subtotal-price">$ {subtotal.toLocaleString()}</span>
        </div>
        <div className="d-flex justify-content-between mb-3">
          <span>Env√≠o:</span>
          <span className="text-success">Gratis</span>
        </div>
        <hr />
        <div className="d-flex justify-content-between mb-3">
          <h5>Total:</h5>
          <h5 className="total-price" id="total-price" style={{ color: 'rgb(4, 129, 20)' }}>
            $ {subtotal.toLocaleString()}
          </h5>
        </div>
        <button className="btn btn-success w-100 py-2" onClick={abrirCheckout}>
          <i className="bi bi-credit-card"></i> PAGAR
        </button>
        <p className="text-muted text-center mt-2 small">
          <i className="bi bi-shield-check"></i> Pago 100% seguro
        </p>
      </div>
    </div>
  </div>
);
