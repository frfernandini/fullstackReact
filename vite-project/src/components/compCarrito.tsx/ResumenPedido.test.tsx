import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, test, vi, beforeEach } from "vitest";
import { ResumenPedido } from "./ResumenPedido";

describe('ResumenPedido', () => {
    let mockAbrirCheckout: ReturnType<typeof vi.fn>;

    beforeEach(() => {
        mockAbrirCheckout = vi.fn();
    });

    test("Debería renderizar correctamente con un subtotal dado", () => {
        const subtotal = 25000;
        render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        // Verificar título
        expect(screen.getByText('Resumen del Pedido')).toBeDefined();
        
        // Verificar subtotal usando getAllByText ya que aparece múltiples veces
        const precioElements = screen.getAllByText(/\$\s*25[.,]000/);
        expect(precioElements.length).toBeGreaterThanOrEqual(1);
        
        // Verificar envío gratis
        expect(screen.getByText('Gratis')).toBeDefined();
        
        // Verificar botón de pagar
        expect(screen.getByText(/PAGAR/i)).toBeDefined();
    });

    test("Debería mostrar el subtotal y el total iguales cuando el envío es gratis", () => {
        const subtotal = 15000;
        render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        // Ambos deben mostrar el mismo valor
        const precioElements = screen.getAllByText(/\$\s*15[.,]000/);
        expect(precioElements.length).toBeGreaterThanOrEqual(2);
    });

    test("Debería llamar a abrirCheckout cuando se hace clic en el botón PAGAR", () => {
        const subtotal = 30000;
        render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        const btnPagar = screen.getByText(/PAGAR/i);
        fireEvent.click(btnPagar);

        expect(mockAbrirCheckout).toHaveBeenCalledTimes(1);
    });

    test("Debería formatear correctamente subtotales grandes con separadores de miles", () => {
        const subtotal = 1250000;
        render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        // Verificar que se formatee con separadores usando getAllByText
        const precioElements = screen.getAllByText(/\$\s*1[.,]250[.,]000/);
        expect(precioElements.length).toBeGreaterThanOrEqual(1);
    });

    test("Debería mostrar el mensaje de pago seguro", () => {
        const subtotal = 10000;
        render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        expect(screen.getByText(/Pago 100% seguro/i)).toBeDefined();
    });

    test("Debería mostrar el ícono de tarjeta de crédito en el botón", () => {
        const subtotal = 10000;
        const { container } = render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        const icon = container.querySelector('.bi-credit-card');
        expect(icon).toBeTruthy();
    });

    test("Debería mostrar el ícono de escudo de seguridad", () => {
        const subtotal = 10000;
        const { container } = render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        const icon = container.querySelector('.bi-shield-check');
        expect(icon).toBeTruthy();
    });

    test("Debería manejar subtotal de 0 correctamente", () => {
        const subtotal = 0;
        render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        // Buscar por el ID específico del subtotal para evitar conflictos
        const subtotalElement = document.getElementById('subtotal-price');
        expect(subtotalElement?.textContent).toMatch(/\$\s*0/);
    });

    test("Snapshot: Debería mantener la estructura del componente", () => {
        const subtotal = 50000;
        const { container } = render(<ResumenPedido subtotal={subtotal} abrirCheckout={mockAbrirCheckout} />);

        expect(container.firstChild).toMatchSnapshot();
    });
});
