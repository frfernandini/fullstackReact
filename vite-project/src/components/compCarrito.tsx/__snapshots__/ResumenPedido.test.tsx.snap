// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResumenPedido > Snapshot: Debería mantener la estructura del componente 1`] = `
<div
  class="order-summary sticky-top"
>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <h4
        class="card-title"
      >
        Resumen del Pedido
      </h4>
      <hr />
      <div
        class="d-flex justify-content-between mb-3"
      >
        <span>
          Subtotal:
        </span>
        <span
          id="subtotal-price"
        >
          $ 
          50,000
        </span>
      </div>
      <div
        class="d-flex justify-content-between mb-3"
      >
        <span>
          Envío:
        </span>
        <span
          class="text-success"
        >
          Gratis
        </span>
      </div>
      <hr />
      <div
        class="d-flex justify-content-between mb-3"
      >
        <h5>
          Total:
        </h5>
        <h5
          class="total-price"
          id="total-price"
          style="color: rgb(4, 129, 20);"
        >
          $ 
          50,000
        </h5>
      </div>
      <button
        class="btn btn-success w-100 py-2"
      >
        <i
          class="bi bi-credit-card"
        />
         PAGAR
      </button>
      <p
        class="text-muted text-center mt-2 small"
      >
        <i
          class="bi bi-shield-check"
        />
         Pago 100% seguro
      </p>
    </div>
  </div>
</div>
`;
